# NBA 2019/2020 Season Restart Prediction

In [40]:
# This Python 3 environment comes with many helpful analytics libraries installed

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import sklearn as sk

In [6]:
fileName_TPGS = 'datasets/2019_2020_Regular_TPGS.csv'
fileName_OPGS = 'datasets/2019_2020_Regular_OPGS.csv'
fileName_Miscellaneous = 'datasets/2019_2020_Miscellaneous.csv'
fileName_Results = 'datasets/2019_2020_Regular_Results.csv'

In [24]:
df_Results = pd.read_csv(fileName_Results)
df_Results.drop(columns = ['Start (ET)', 'Unnamed: 6', 'Unnamed: 7', 'Notes', 'Attend.', 'Notes'], inplace = True)
df_Results["MarginHome"] = df_Results['PTS.1'] - df_Results['PTS']
df_Results.columns = ['Date', 'Team Away', 'Points Away', 'Team Home', 'Points Home', 'Margin Home']

In [25]:
df_Results

Date             Team Away  Points Away  \
0    Tue Oct 22 2019  New Orleans Pelicans          122   
1    Tue Oct 22 2019    Los Angeles Lakers          102   
2    Wed Oct 23 2019         Chicago Bulls          125   
3    Wed Oct 23 2019       Detroit Pistons          119   
4    Wed Oct 23 2019   Cleveland Cavaliers           85   
..               ...                   ...          ...   
966  Tue Mar 10 2020  Los Angeles Clippers          131   
967  Wed Mar 11 2020       Detroit Pistons          106   
968  Wed Mar 11 2020       New York Knicks          136   
969  Wed Mar 11 2020     Charlotte Hornets          109   
970  Wed Mar 11 2020        Denver Nuggets           97   

                 Team Home  Points Home  Margin Home  
0          Toronto Raptors          130            8  
1     Los Angeles Clippers          112           10  
2        Charlotte Hornets          126            1  
3           Indiana Pacers          110           -9  
4            Orlando Magic           94            9  
..                     ...          ...          ...  
966  Golden State Warriors          107          -24  
967     Philadelphia 76ers          124           18  
968          Atlanta Hawks          131           -5  
969             Miami Heat           98          -11  
970       Dallas Mavericks          113           16  

[971 rows x 6 columns]

## How frequently do the home teams win?

Typically, the assumption is that home teams will win more often than not.

In [27]:
df_Results['Home Win'] = df_Results['Margin Home'] > 0
y_true = df_Results['Home Win'].values

In [28]:
df_Results

Date             Team Away  Points Away  \
0    Tue Oct 22 2019  New Orleans Pelicans          122   
1    Tue Oct 22 2019    Los Angeles Lakers          102   
2    Wed Oct 23 2019         Chicago Bulls          125   
3    Wed Oct 23 2019       Detroit Pistons          119   
4    Wed Oct 23 2019   Cleveland Cavaliers           85   
..               ...                   ...          ...   
966  Tue Mar 10 2020  Los Angeles Clippers          131   
967  Wed Mar 11 2020       Detroit Pistons          106   
968  Wed Mar 11 2020       New York Knicks          136   
969  Wed Mar 11 2020     Charlotte Hornets          109   
970  Wed Mar 11 2020        Denver Nuggets           97   

                 Team Home  Points Home  Margin Home  Home Win  
0          Toronto Raptors          130            8      True  
1     Los Angeles Clippers          112           10      True  
2        Charlotte Hornets          126            1      True  
3           Indiana Pacers          110           -9     False  
4            Orlando Magic           94            9      True  
..                     ...          ...          ...       ...  
966  Golden State Warriors          107          -24     False  
967     Philadelphia 76ers          124           18      True  
968          Atlanta Hawks          131           -5     False  
969             Miami Heat           98          -11     False  
970       Dallas Mavericks          113           16      True  

[971 rows x 7 columns]

In [29]:
total_games = df_Results['Home Win'].count()
total_home_wins = df_Results['Home Win'].sum()
win_percent = total_home_wins / total_games

print('Home Win Percentage: {0:.2f}%'.format(100 * win_percent))

Home Win Percentage: 55.10%


The home teams' win percentage is 58.0%. If a model is considered useful, it must (at minimum) have a better score than the established baseline.

### Setting up a Classifier

In [31]:
# Predicting the baseline for Home Teams wins with Binary Classification

from sklearn.metrics import f1_score

y_pred = [1] * len(y_true)
print('F1: {0:.4f}%'.format(f1_score(y_true, y_pred, pos_label = None, average = 'weighted') * 100))

F1: 39.1465%


The calculated F1 score of 39.1% is the score to beat.

## Streaks

In [41]:
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(random_state = 2020)

In [42]:
df_Results['Home Last Win'] = False
df_Results['Visitor Last Win'] = False